app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx

```diff
@@ -1,48 +1,232 @@
+      setTimeout(() => isMountedRef.current && formik.setFieldValue('variables', [{ key: null, value: null }]), 100);
+    }
+  }, []);
+
+  const FieldLayoutComponent = React.useContext(LayoutContext);
+
+  const [deleteCount, setDeleteCount] = React.useState(0);
+
+  return (
+    <FieldArray
+      key={deleteCount}
+      name="variables"
+      render={arrayHelpers => (
+        <div className="EvaluateVariablesStageConfig form-horizontal">
+          <FieldLayoutComponent
+            label={<h4>Variable Name</h4>}
+            input={<h4>Variable Value</h4>}
+            validation={{ hidden: true } as any}
           />
-        </StageConfigField>
-      </div>
-    );
-  }
+
+          {variables.map((_, index) => {
+            const onDeleteClicked = () => {
+              setDeleteCount(count => count + 1);
+              arrayHelpers.handleRemove(index)();
+            };
+
+            return (
+              <FormikVariable
+                key={`${deleteCount}-${index}`}
+                index={index}
+                previewStage={chosenStage}
+                onDeleteClicked={onDeleteClicked}
+              />
+            );
+          })}
+
+          <button
+            type="button"
+            className="btn btn-block btn-sm add-new"
+            onClick={arrayHelpers.handlePush({ key: null, value: null })}
+          >
+            <span className="glyphicon glyphicon-plus-sign" />
+            Add Variable
+          </button>
+        </div>
+      )}
+    />
+  );
+}
+
+interface IFormikVariableProps {
+  index: number;
+  onDeleteClicked(): void;
+  previewStage: IStageForSpelPreview;
+}
+
+function FormikVariable({ index, onDeleteClicked, previewStage }: IFormikVariableProps) {
+  const [touchedOverride, setTouchedOverride] = React.useState(undefined);
+
+  const variableNameInputAsLabel = (
+    <FormikFormField
+      name={`variables[${index}].key`}
+      required={true}
+      input={inputProps => <TextInput {...inputProps} placeholder="Variable name" />}
+      layout={VariableNameFormLayout}
+    />
+  );
+
+  const actions = (
+    <Tooltip value="Remove variable">
+      <button className="btn btn-sm btn-default" onClick={onDeleteClicked}>
+        <span className="glyphicon glyphicon-trash" />
+      </button>
+    </Tooltip>
+  );
+
+  const fieldName = `variables[${index}].value`;
+
+  return (
+    <FormikFormField
+      name={fieldName}
+      onChange={value => {
+        // When the user has entered anything, mark the field as touched so warnings are shown
+        if (value && !touchedOverride) {
+          setTouchedOverride(true);
+        }
+      }}
+      touched={touchedOverride}
+      label={variableNameInputAsLabel}
+      actions={actions}
+      fastField={false}
+      input={inputProps => (
+        <SpelInput
+          {...inputProps}
+          placeholder="Variable value, e.g. ${trigger.properties.myproperty}"
```

Comment:
nit: i would prefer a better/more real example, eg.
```suggestion
          placeholder="Variable value, e.g. ${trigger.buildInfo.number}"
```